packages/artifact_proxy/bin/artifact_proxy.dart

```diff
@@ -0,0 +1,128 @@
+  // Don't watch the file if we're also writing to it!
+  if (shouldRecord) {
+    return;
+  }
+  final watcher = FileWatcher(configPath);
+  watcher.ready.then((_) {
+    print('Watching ${watcher.path}');
+  });
+  watcher.events.listen((event) async {
+    print('Config changed, restarting server');
+    server.close();
+    server = await createServer(configPath, shouldRecord);
+  });
+}
+
+void recordArtifact(String configPath, String path) {
+  final edit = YamlEditor(File(configPath).readAsStringSync());
+  var existing = edit.parseAt(['known_redirects'], orElse: () => YamlList());
+  if (existing is! YamlList || existing.isEmpty) {
+    edit.update(['known_redirects'], [path]);
+    File(configPath).writeAsStringSync(edit.toString());
+  } else if (!existing.contains(path)) {
+    edit.appendToList(['known_redirects'], path);
+    File(configPath).writeAsStringSync(edit.toString());
+  }
+  // Already contains the key, do nothing.
+}
+
+Future<HttpServer> createServer(String configPath, bool shouldRecord) async {
+  final configYaml = loadYaml(File(configPath).readAsStringSync());
+  final shorebirdEngine = configYaml['shorebird_engine'];
+  final flutterEngine = configYaml['flutter_engine'];
+
+  print("Mapping $shorebirdEngine to $flutterEngine");
+
+  // This doesn't work yet, this was from a previous attempt before I
+  // started down a config.yaml path.
+  const knownRedirects = <String>[
+    'sky_engine.zip',
+    'fonts.zip',
+    'flutter_patched_sdk.zip',
+    'flutter_patched_sdk_product.zip',
+    'android-arm-profile/darwin-x64.zip',
+    'android-arm-release/darwin-x64.zip',
+    'android-arm64-profile/darwin-x64.zip',
+    'android-arm64-release/darwin-x64.zip',
+    'android-x64-profile/darwin-x64.zip',
+    'android-x64-release/darwin-x64.zip',
+    'darwin-arm64/artifacts.zip',
+    'darwin-arm64/font-subset.zip',
+    'dart-sdk-darwin-arm64.zip',
+    'gradle-wrapper.tgz',
+  ];
+
+  bool isKnownRedirect(Uri url) {
+    for (var knownPath in knownRedirects) {
+      if (url.path.endsWith(knownPath)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  String redirectUrl(Uri url) {
+    final path = url.path;
+    if (path.contains('+')) {
+      // Assume this is a cipd URL rather than a Google Storage URL
+      // and undo the cipd path mangling.
+      // https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/cache.dart#L475
+      return path.replaceAll('flutter_infra_release/cipd',
+          'https://chrome-infra-packages.appspot.com/dl');
+    }
+    // Otherwise assume it's a Google Storage URL but redirect from our
+    // Flutter engine version to the base Flutter engine version.
+    var adjustedPath = path.replaceAll(shorebirdEngine, flutterEngine);
+    return 'https://storage.googleapis.com/$adjustedPath';
+  }
+
+  var handler = const Pipeline()
+      .addMiddleware(logRequests())
+      .addHandler((Request request) {
+    if (shouldRecord) {
```

Comment:
Minor but using [shelf_proxy](https://pub.dev/packages/shelf_proxy) might come in handy (doesn't seem to matter too much since there's not too much logic but if this grow it could become simpler to have multiple proxy handlers for different resources/locations).